api/src/database/repositories/ThreadRepository.ts

```diff
@@ -75,38 +76,52 @@ export const ThreadRepository = AppDataSource.getRepository(Thread).extend({
     return users;
   },
 
-  async searchThreads(searchText: string, projectId: number): Promise<Thread[]> {
-    const formattedSearchText = searchText.split(' ').join('&');
+  async searchThreads(searchText: string, projectId: number, limit: number = 10): Promise<Thread[]> {
+    const targetVector: number[] = await createEmbedding(searchText);
+    const targetVectorString = targetVector.join(',');
+
+    const threads = await this.createQueryBuilder('thread')
+      .addSelect(`embedding::vector <-> ARRAY[${targetVectorString}]::vector`, 'similarity')
+      .where('thread.projectId = :projectId', { projectId })
+      .orderBy('similarity', 'ASC')
+      .limit(limit)
+      .getMany();
 
+    const threadIds = threads.map((thread) => thread.id);
+    const threadOrder = threadIds.join(',');
+
+    return await this.getThreadsWithProperties()
+      .where('thread.projectId = :projectId', { projectId })
+      .andWhereInIds(threadIds)
+      .orderBy(`ARRAY_POSITION(ARRAY[${threadOrder}], thread.id)`)
+      .getMany();
+  },
+
+  getThreadsWithProperties(): SelectQueryBuilder<Thread> {
     return this.createQueryBuilder('thread')
-      .leftJoinAndSelect('thread.snippets', 'snippet')
+      .leftJoinAndSelect('thread.snippets', 'threadSnippets')
       .leftJoinAndSelect('thread.user', 'user')
       .leftJoinAndSelect('thread.replies', 'reply')
-      .where(`thread.projectId = :projectId`, { projectId })
-      .andWhere(
-        new Brackets((qb) => {
-          qb.where(`to_tsvector('english', thread.title) @@ to_tsquery('english', :formattedSearchText)`, {
-            formattedSearchText,
-          })
-            .orWhere(`to_tsvector('english', thread.message) @@ to_tsquery('english', :formattedSearchText)`, {
-              formattedSearchText,
-            })
-            .orWhere(`to_tsvector('english', reply.message) @@ to_tsquery('english', :formattedSearchText)`, {
-              formattedSearchText,
-            })
-            .orWhere(`to_tsvector('english', snippet.text) @@ to_tsquery('english', :formattedSearchText)`, {
-              formattedSearchText,
-            });
-        })
-      )
-      .orderBy(`ts_rank(to_tsvector('english', thread.title), to_tsquery('english', :formattedSearchText))`, 'DESC')
-      .addOrderBy(
-        `ts_rank(to_tsvector('english', thread.message), to_tsquery('english', :formattedSearchText))`,
-        'DESC'
-      )
-      .addOrderBy(`ts_rank(to_tsvector('english', reply.message), to_tsquery('english', :formattedSearchText))`, 'DESC')
-      .addOrderBy(`ts_rank(to_tsvector('english', snippet.text), to_tsquery('english', :formattedSearchText))`, 'DESC')
-      .loadRelationCountAndMap('thread.replyCount', 'thread.replies')
-      .getMany();
+      .leftJoinAndSelect('reply.snippets', 'replySnippets');
+  },
+
+  async updateEmbeddingsOfAllThreads(projectId: number): Promise<void> {
```

Comment:
i don't see this function being used anywhere, what's the purpose of it?

Reply:
Its right on the method above, on threadController, and replyController.

The purpose of it is update the embedding when anything under a thread is modified.

Reply:
threadController is using the `updateSearchEmbeddingsForThread` func. I don't see the `updateEmbeddingsOfAllThreads` being used in that file.